marlowe-chain-sync/libchainsync/Language/Marlowe/Runtime/ChainSync/QueryServer.hs

```diff
@@ -58,49 +55,46 @@ chainSyncQueryServer ChainSyncQueryServerDependencies{..} = ServerSource $ pure
+              GetEra -> either (fail . show) pure =<< queryLocalNodeState Nothing QueryCurrentEra
         }
 
-    queryGenesisParameters :: (GenesisParameters -> a) -> m a
-    queryGenesisParameters f = f <$> queryShelley (const QueryGenesisParameters)
+    queryGenesisParameters :: (forall era. GenesisParameters era -> a) -> m a
+    queryGenesisParameters f = f <$> queryShelley (const $ QueryInShelleyBasedEraProjection QueryGenesisParameters id)
 
     queryShelley
-      :: (forall era. ShelleyBasedEra era -> QueryInShelleyBasedEra era a)
-      -> m a
-    queryShelley query =
+      :: forall r
+       . (forall era. ShelleyBasedEra era -> QueryInShelleyBasedEraProjection era r)
+      -> m r
+    queryShelley mkQuery =
       either fail pure =<< runExceptT do
-        AnyCardanoEra era <-
-          withExceptT show $
-            ExceptT $
-              queryLocalNodeState Nothing $
-                QueryCurrentEra CardanoModeIsMultiEra
-        eraInMode <- case toEraInMode era CardanoMode of
-          Nothing -> throwE $ "cannot convert " <> show era <> " to era in mode"
-          Just eraInMode -> pure eraInMode
-        shelleyBasedEra <- case eraInMode of
-          ByronEraInCardanoMode -> throwE "Cannot query shelley in byron era"
-          ShelleyEraInCardanoMode -> pure ShelleyBasedEraShelley
-          AllegraEraInCardanoMode -> pure ShelleyBasedEraAllegra
-          MaryEraInCardanoMode -> pure ShelleyBasedEraMary
-          AlonzoEraInCardanoMode -> pure ShelleyBasedEraAlonzo
-          BabbageEraInCardanoMode -> pure ShelleyBasedEraBabbage
-          ConwayEraInCardanoMode -> pure ShelleyBasedEraConway
-        result <-
-          withExceptT show $
-            ExceptT $
-              queryLocalNodeState Nothing $
-                QueryInEra eraInMode $
-                  QueryInShelleyBasedEra shelleyBasedEra $
-                    query shelleyBasedEra
-        withExceptT show $ except result
+        AnyCardanoEra era <- withExceptT show $ ExceptT $ queryLocalNodeState Nothing QueryCurrentEra
+        shelleyBasedEra <- inEonForEra (throwE "Cannot query shelley in byron era") pure era
+        withCurrentEra shelleyBasedEra
+      where
+        withCurrentEra :: forall era. ShelleyBasedEra era -> ExceptT String m r
+        withCurrentEra shelleyBasedEra = case mkQuery shelleyBasedEra of
+          QueryInShelleyBasedEraProjection query projection -> do
+            result <-
+              withExceptT show
+                . ExceptT
+                . queryLocalNodeState Nothing
+                . QueryInEra
+                $ QueryInShelleyBasedEra shelleyBasedEra query
+            withExceptT show $ except $ projection <$> result
 
     traverseReqTree :: (forall x. ChainSyncQuery x -> m x) -> ReqTree ChainSyncQuery a -> m a
     traverseReqTree f = \case
       ReqLeaf req -> f req
       ReqBin l r -> (,) <$> traverseReqTree f l <*> traverseReqTree f r
+
+data QueryInShelleyBasedEraProjection era r where
```

Comment:
Nice!

Reply:
Ah, yeah I wish I didn't have to do this :laughing: